### DISCUSSION TITLE
How are the SYCL  Atomic operations implemented on ROCM architectures?

### DISCUSSION THREAD

**Discussion by csccva** [AdaptiveCpp/AdaptiveCpp#1638] [General]
Hello,

I found out the hard way that in hip/rocm the`atomicAdd` function  is very slow unless using the flag `-munsafe-fp-atomics ` or using the alternative function `unsafeAtomicAdd`. 

There is no equivalent in CUDA. It is always equivalent to the "unsafe" atomics. 

Which version does ACPP and SYCL in general uses on rocm architectures?

Cristian

**Comment by illuhad** [comment]
`unsafeAtomicAdd` only makes a difference on `gfx90a`. On all other architectures I've looked at it results in a regular `atomicrmw` instruction:
https://godbolt.org/z/TTYnx69jx

We can hardly use something that is labelled as "unsafe" by default ;)

* `--acpp-targets=hip` behaves the same as (I think) `hipcc`. It does not use unsafe atomic add by default, but you can enable it using `-munsafe-fp-atomics`.
* `generic` JIT compiler does  not use unsafe atomics. We could fairly easily expose something equivalent to `unsafeAtomicAdd` that users could call when JIT-compiling for AMD specifically.

> There is no equivalent in CUDA. It is always equivalent to the "unsafe" atomics.

I'm not sure this is true. I think atomic operations in CUDA always produce "correct" results.

**Comment by csccva** [comment]
The "unsafe" term does not mean that the results are wrong. Someone from AMD explained to me that doing atomics on a variable in the GPU memory will give the "correct" results when using the unsafe version. I did not get the explanation about the coarse vs. finee - grained memory, but I saw this  (https://docs.olcf.ornl.gov/systems/frontier_user_guide.html#fp-atomic-ops-coarse-fine-allocations) 
 

```
Users applying floating point atomic operations (e.g., atomicAdd) on memory regions allocated via regular hipMalloc() can safely apply the -munsafe-fp-atomics flags to their codes to get the best possible performance and leverage hardware supported floating point atomics. 
```

I am working on a port from serial cpu fortran code to hip and I saw no difference in the results beyond rounding errors between the two. My kernel that was using (more like abusing) atomics got a 100 speed up by using `munsafe-atomics` flag.

Also as I understand from https://rocm.docs.amd.com/projects/HIP/en/latest/reference/cpp_language_extensions.html the "safe" atomics are software emulations using  compare-and-swap (CAS) loop, while in CUDA as far as I understand the builtin atomicxxx functions are hardware implementations similar to the "unsafe" AMS atomics. 

Some more info in the presentation page 28. Though I do not understand everything I think that at least for the MI250X gpus this is important issue if the code uses atomics.

Good to know that I can add the `-munsafe-fp-atomics` to the acpp. I wonder if OneAPI has something similar.

Cristian

**Reply by illuhad** [reply]
> The "unsafe" term does not mean that the results are wrong. Someone from AMD explained to me that doing atomics on a variable in the GPU memory will give the "correct" results when using the unsafe version. I did not get the explanation about the coarse vs. finee - grained memory, but I saw this (https://docs.olcf.ornl.gov/systems/frontier_user_guide.html#fp-atomic-ops-coarse-fine-allocations)

Their own documentation explicitly states that results may be numerically different or incorrect in certain cases.
https://rocm.docs.amd.com/projects/HIP/en/latest/reference/cpp_language_extensions.html
And even if perhaps current hardware is more well-behaved, future hardware might be more erratic here given how they have defined this functionality.

From their documentation:
> On some devices, fast atomic RMW instructions can produce results that differ from the same functions implemented with atomic CAS loops. For example, some devices will use different rounding or denormal modes, and some devices produce incorrect answers if fast floating-point atomic RMW instructions target fine-grained memory allocations.

The fine-grained case probably refers to what we in SYCL call `sycl::malloc_shared`, which is something that can always be involved (which the compiler also cannot know). So we absolutely cannot use unsafe atomics by default.

My guess, since this only affects gfx90a (MI2xx) and not later architectures like gfx942 (MI300) is that they messed something up in hardware with atomics not providing the guarantees actually needed by software, so software people had to find some workaround, which was then fixed in hardware for later generations.

**Reply by csccva** [reply]
I am not at the level to recommend something like that.  As long as I can use the `munsafe-fp-atomics`, for me  is ok. 
Thanks for the explanation.
